### PR TITLE
Improve `new Vector2/3/4(c1, c2, c3, c4)` codegen

### LINKED ISSUE #17256: Constructing a `Vector4` from 4 constant floats is very inefficient
Code that is in the form of: `var value = new Vector4(1, 2, 3, 4)` currently generates very inefficient code.

Today, this produces something similar to:
```asm
C4E17A100518010000   vmovss   xmm0, dword ptr [reloc @RWD00]
C4E17A100D13010000   vmovss   xmm1, dword ptr [reloc @RWD04]
C4E17A10150E010000   vmovss   xmm2, dword ptr [reloc @RWD08]
C4E17A101D09010000   vmovss   xmm3, dword ptr [reloc @RWD12]
C4E15857E4           vxorps   xmm4, xmm4
C4E15A10E3           vmovss   xmm4, xmm4, xmm3
C4E15973FC04         vpslldq  xmm4, 4
C4E15A10E2           vmovss   xmm4, xmm4, xmm2
C4E15973FC04         vpslldq  xmm4, 4
C4E15A10E1           vmovss   xmm4, xmm4, xmm1
C4E15973FC04         vpslldq  xmm4, 4
C4E15A10E0           vmovss   xmm4, xmm4, xmm0
C4E17828C4           vmovaps  xmm0, xmm4
```

The JIT should recognize this type of code and instead optimize it to be a single read:
```asm
vmovaps xmm0 xmm ptr [address]    ; Use vmovups if address is not aligned
```

category:cq
theme:vector-codegen
skill-level:intermediate
cost:medium

### PR DESCRIPTION
**Pull request by mikedn**
Fixes #17256

### CODE REVIEW

**Comment by mikedn** [comment]
jit-diff summary:
```
Total bytes of diff: -546 (0.00% of base)
    diff is an improvement.
Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes
Top file improvements by size (bytes):
        -334 : System.Numerics.Vectors.dasm (-0.59% of base)
        -101 : System.Runtime.Numerics.dasm (-0.17% of base)
         -59 : System.Private.CoreLib.dasm (0.00% of base)
         -34 : Microsoft.VisualBasic.dasm (-0.02% of base)
          -8 : System.Linq.Expressions.dasm (0.00% of base)
7 total files with size differences (7 improved, 0 regressed), 72 unchanged.
Top method regessions by size (bytes):
          47 : System.Numerics.Vectors.dasm - Matrix4x4:Invert(struct,byref):bool
Top method improvements by size (bytes):
         -64 : System.Numerics.Vectors.dasm - Matrix4x4:Negate(struct):struct
         -64 : System.Numerics.Vectors.dasm - Matrix4x4:op_UnaryNegation(struct):struct
         -24 : System.Numerics.Vectors.dasm - Matrix3x2:Negate(struct):struct
         -24 : System.Numerics.Vectors.dasm - Matrix3x2:op_UnaryNegation(struct):struct
         -20 : System.Private.CoreLib.dasm - MathF:IEEERemainder(float,float):float
57 total methods with size differences (56 improved, 1 regressed), 66535 unchanged.
```

**Comment by mikedn** [comment]
@dotnet/jit-contrib Do you want this? Needs a bit of work before it reaches a mergeable state...

Anyone knows if there are any gotchas associated with enabling 16 byte aligned data constants in the emitter?

**Comment by tannergooding** [comment]
@mikedn, if you do enable 16-byte aligned data constants, I believe you can also fix: https://github.com/dotnet/coreclr/blob/76dcdad7c77705e99b4077d81cfd8f22c3d01711/src/jit/lsraxarch.cpp#L1930

**Comment by mikedn** [comment]
Of all open PRs I have this one is a bit likely to bite the dust. While it does improve code size it does so at the cost of increasing the data section size. 

This optimization is a no-brainer in native compilers that can share such constants across functions but in a JIT where each function gets its own data section this is a bit questionable. At the very least I'd like the JIT to dump these data sections so that they show up in diffs (I have work done for that but it is not pushed yet). 

It would also be useful to find a way to minimize the impact of the rather large alignment required by such constants, right now I'm not sure what can be done about that.

**Comment by CarolEidt** [comment]
Thanks for the update. I'll set this one as "Future", and you can decide whether to close or leave it open.

**Comment by BruceForstall** [comment]
Maybe the JIT could generate pools in a crossgen or AOT mode. Or if some metric like number of use instances warrants it.

> At the very least I'd like the JIT to dump these data sections so that they show up in diffs (I have work done for that but it is not pushed yet).

It we don't do that already, then by all means we should dump them!

**Comment by mikedn** [comment]
Well, I don't know if we'll end up doing NEG/ABS but the data section improvements are good for #17256 and that one really should be fixed, it generates way too much code. So I'll keep working on this.

**Comment by mikedn** [comment]
@CarolEidt What do you think about lowering `SIMDIntrinsicInitN(c1, c2, c3, c4)` to `IND(CLS_VAR_ADDR(dataSectionHandle))` to solve #17256? See the last commit above.

It looks like the JIT never attempted to create data sections before codegen but there doesn't appear to be anything preventing us from doing that. The alternative is to mark all 4 arguments as contained but that requires more code changes (`ContainCheckSIMD`, `BuildSIMD` and `genSIMDIntrinsicInitN`). It's also not clear if we can then make the constant vector contained like we could probably contain the indir.

`return new Vector4(1, 2, 3, 4) + new Vector4(4, 3, 2, 1);` generates 

```asm
G_M33474_IG01:
       C5F877               vzeroupper
G_M33474_IG02:
       C4E179100524000000   vmovupd  xmm0, xmmword ptr [reloc @RWD00]
       C4E179100D2B000000   vmovupd  xmm1, xmmword ptr [reloc @RWD16]
       C4E17858C1           vaddps   xmm0, xmm1 ; could be contained
       C4E1791101           vmovupd  xmmword ptr [rcx], xmm0
       488BC1               mov      rax, rcx
G_M33474_IG03:
       C3                   ret

RWD00  db       000h, 000h, 080h, 03Fh, 000h, 000h, 000h, 040h, 000h, 000h, 040h, 040h, 000h, 000h, 080h, 040h
RWD16  db       000h, 000h, 080h, 040h, 000h, 000h, 040h, 040h, 000h, 000h, 000h, 040h, 000h, 000h, 080h, 03Fh
```

**Comment by CarolEidt** [comment]
> @CarolEidt What do you think about lowering SIMDIntrinsicInitN(c1, c2, c3, c4) to IND(CLS_VAR_ADDR(dataSectionHandle)) to solve #17256? See the last commit above.

I like it - I think it's a good approach; I wouldn't have thought it would be this straightforward to do this - thanks (and sorry for the long delay in response).

File: src/jit/emit.cpp [hunk 1]
```diff
@@ -5189,20 +5188,27 @@ UNATIVE_OFFSET emitter::emitDataGenBeg(UNATIVE_OFFSET size, bool dblAlign, bool
...
-    if (dblAlign && (secOffs % sizeof(double) != 0))
+    if (align)
     {
-        /* Need to skip 4 bytes to honor dblAlign */
-        /* Must allocate a dummy 4 byte integer */
-        int zero = 0;
-        emitDataGenBeg(4, false, false);
-        emitDataGenData(0, &zero, 4);
-        emitDataGenEnd();
+        assert(size <= 16);
```
Comment (CarolEidt) [inline src/jit/emit.cpp]: Wouldn't this assert hold for any case (not just when `align` is true)?

File: src/jit/emit.cpp [hunk 2]
```diff
@@ -5533,6 +5562,111 @@ void emitter::emitOutputDataSec(dataSecDsc* sec, BYTE* dst)
     }
 }
 
+#ifdef DEBUG
+void emitter::emitDispDataSec(dataSecDsc* section)
```
Comment (CarolEidt) [inline src/jit/emit.cpp]: Nice! Might I ask for a basic function header?

File: src/jit/emit.cpp [hunk 3]
```diff
@@ -5533,6 +5562,111 @@ void emitter::emitOutputDataSec(dataSecDsc* sec, BYTE* dst)
...
+
+                if (isRelative)
+                {
+                    if (emitComp->opts.disDiffable)
+                    {
+                        printf("dd\t%s - %s\n", blockLabel, firstLabel);
+                    }
+                    else
+                    {
+                        printf("dd\t%08Xh", ig->igOffs - igFirst->igOffs);
```
Comment (CarolEidt) [inline src/jit/emit.cpp]: Even in non-diffable format would it be useful to print the labels as well?

**Comment by mikedn** [comment]
>  I wouldn't have thought it would be this straightforward to do this - thanks (and sorry for the long delay in response).

Well, it would be better if the constants were sorted by size to minimize alignment holes. There's also no support for 32 byte aligned constants, the VM lacks an option for that.

No problem about the delay, there's plenty of things to do in the JIT world and this change doesn't block anything else :)

File: src/jit/emit.cpp (hunk 2 above)
Comment (mikedn) [reply src/jit/emit.cpp]: Sure, will add one.

File: src/jit/emit.cpp (hunk 1 above)
Comment (mikedn) [reply src/jit/emit.cpp]: Constants can have any size (e.g. 32 bytes for AVX stuff), it's just that alignment is currently supported only for 4, 8 and 16 bytes constants. Hmm, maybe I should check those specific sizes instead

File: src/jit/emit.cpp (hunk 3 above)
Comment (mikedn) [reply src/jit/emit.cpp]: Hah, I forgot that it does something like this already. Here's the diffable output:
```
RWD00  dd       G_M55886_IG03 - G_M55886_IG02
       dd       G_M55886_IG05 - G_M55886_IG02
       dd       G_M55886_IG07 - G_M55886_IG02
       dd       G_M55886_IG09 - G_M55886_IG02
       dd       G_M55886_IG10 - G_M55886_IG02
```
and the non-diffable output:
```
RWD00  dd       00000023h ; case G_M55886_IG03
       dd       00000029h ; case G_M55886_IG05
       dd       0000002Fh ; case G_M55886_IG07
       dd       00000035h ; case G_M55886_IG09
       dd       0000003Ch ; case G_M55886_IG10
```

**Comment by mikedn** [comment]
> it would be better if the constants were sorted by size to minimize alignment holes

Actually they are. As is now all 16 byte constants are created in lowering and all other constants (4/8 bytes in size) are created during codegen. So all 16 byte constants come first and there are no alignment holes between them.

This will only be a problem if we try to extend it to 32 byte constants for AVX intrinsics. But for `Vector4` this works just fine.

**Comment by mikedn** [comment]
This works well even for `Vector2` where one may expect that building the vector from components isn't that bad. It kind of is:

```asm
; before
...
       vmovss   xmm0, dword ptr [reloc @RWD00]
       vmovss   xmm1, dword ptr [reloc @RWD04]
       vxorps   xmm2, xmm2
       vmovss   xmm2, xmm2, xmm1
       vpslldq  xmm2, 4
       vmovss   xmm2, xmm2, xmm0
       vmovaps  xmm0, xmm2
       vmovsd   qword ptr [rsp+10H], xmm0
       vmovss   xmm0, dword ptr [reloc @RWD08]
       vmovss   xmm1, dword ptr [reloc @RWD12]
       vxorps   xmm2, xmm2
       vmovss   xmm2, xmm2, xmm1
       vpslldq  xmm2, 4
       vmovss   xmm2, xmm2, xmm0
       vmovaps  xmm0, xmm2
       vmovsd   qword ptr [rsp+08H], xmm0
       vmovsd   xmm0, qword ptr [rsp+10H]
       vmovsd   xmm1, qword ptr [rsp+08H]
       vaddps   xmm0, xmm1
       vmovd    rax, xmm0
...

RWD00  dd       3F800000h
RWD04  dd       40000000h
RWD08  dd       40800000h
RWD12  dd       40400000h
```

```asm
; after
...
       vmovsd   xmm0, qword ptr [reloc @RWD00]
       vmovsd   qword ptr [rsp+10H], xmm0
       vmovsd   xmm0, qword ptr [reloc @RWD16]
       vmovsd   qword ptr [rsp+08H], xmm0
       vmovsd   xmm0, qword ptr [rsp+10H]
       vmovsd   xmm1, qword ptr [rsp+08H]
       vaddps   xmm0, xmm1
       vmovd    rax, xmm0
...
RWD00  db       000h, 000h, 080h, 03Fh, 000h, 000h, 000h, 040h, 000h, 000h, 000h, 000h, 000h, 000h, 000h, 000h
RWD16  db       000h, 000h, 080h, 040h, 000h, 000h, 040h, 040h, 000h, 000h, 000h, 000h, 000h, 000h, 000h, 000h
```
Uses 16 bytes more of data but saves 70 bytes of code (138 bytes vs. 68 bytes).

File: src/jit/emit.cpp (hunk 1 above)
Comment (mikedn) [reply src/jit/emit.cpp]: I added a comment to clarify this assert.

**Comment by CarolEidt** [comment]
@dotnet-bot test Ubuntu x64 Checked CoreFX Test
@dotnet-bot test Ubuntu x64 Checked Innerloop Build and Test
@dotnet-bot test Ubuntu x64 Formattin

**Comment by 4creators** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test 
@dotnet-bot test Ubuntu x64 Checked CoreFX Test
@dotnet-bot test Ubuntu x64 Formatting

**Comment by 4creators** [comment]
@dotnet-bot test Ubuntu x64 Checked CoreFX Test

**Comment by 4creators** [comment]
@dotnet-bot test Ubuntu x64 Checked CoreFX Test

**Comment by 4creators** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test
@dotnet-bot test Ubuntu x64 Checked CoreFX Test

**Comment by 4creators** [comment]
@dotnet-bot test Ubuntu x64 Checked CoreFX Test

**Comment by mikedn** [comment]
@dotnet-bot test Ubuntu x64 Checked CoreFX Tests

**Comment by mikedn** [comment]
The WIP both got stuck, it didn't update the WIP check even after changing the title.

**Comment by tannergooding** [comment]
Has there been any consideration on optimizing the case where most (but not all) of the inputs are constant?

For example: `Create(c, c, v, c)` (where c is a constant and v is a variable input)? I would guess that it would be better to create a 16-byte vector constant where the variable argument is zero and then have codegen do a single insertion (similar to what the C# compiler does for the arrays that are mostly constant).

**Comment by mikedn** [comment]
> For example: Create(c, c, v, c) (where c is a constant and v is a variable input)? I would guess that it would be better to create a 16-byte vector constant where the variable argument is zero and then have codegen do a single insertion (similar to what the C# compiler does for the arrays that are mostly constant).

Yes, that's doable. It's just that it's a lot of work - looking at various interesting patterns, figuring out the right trade offs, deciding what instructions to generate depending on the number of variables to insert, their position, the available ISA etc. I've looked a bit at what native compilers do here and it seems pretty daunting to do something comprehensive.

Just an example: For
```C++
_mm256_set_epi32(x, 1, x, 1, x, 1, x, 1);
```
VC++ generates
```asm
vmovdqu xmm0, XMMWORD PTR __xmm@00000000000000010000000000000001
vpinsrd xmm0, xmm0, ecx, 1
vpinsrd xmm0, xmm0, ecx, 3
vinsertf128 ymm0, ymm0, xmm0, 1
```

**Comment by 4creators** [comment]
@BruceForstall @mikedn Any chance in having this PR reviewed and merged before 3.0 ships?

**Comment by fiigii** [comment]
@CarolEidt @mikedn Is this PR ready to merge? I want to make the similar optimizations for `Vector128/256` as well.

**Comment by mikedn** [comment]
This one is done but I'm considering using a somewhat different approach in the future.

We should probably have a `GT_CNS_SIMD` node and handle that the same way floating point constants are handled. That would help in other cases (e.g. initializing a block with a non-zero bit pattern, avoid the need for special codegen for `NI_SSE41_TestAllOnes`) and perhaps would allow reusing constants already loaded into registers via LSRA's constant intervals.

But that's probably quite a bit of work and thus seems unlikely to catch .NET 3.0. So for now this is probably the best we can do.

File: src/jit/emit.h [hunk 4]
```diff
@@ -216,6 +216,15 @@ class emitLocation
     unsigned  codePos; // the code position within the IG (see emitCurOffset())
 };
 
+#ifndef LEGACY_BACKEND
```
[the remark this answers is not in the document]
Comment (mikedn) [reply src/jit/emit.h]: Yes, thanks!

**Comment by mikedn** [comment]
OSX failed due to the usual Jenkins exception:
```
00:32:07.313 Test passed.
00:32:07.338 ...FATAL: command execution failed
00:32:21.662 java.nio.channels.ClosedChannelException
00:32:21.662 	at org.jenkinsci.remoting.protocol.NetworkLayer.onRecvClosed(NetworkLayer.java:154)
00:32:21.662 	at org.jenkinsci.remoting.protocol.impl.NIONetworkLayer.ready(NIONetworkLayer.java:142)
00:32:21.662 	at org.jenkinsci.remoting.protocol.IOHub$OnReady.run(IOHub.java:721)
```
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Pull request merged**